src/components/fields/ArrayField.js

```diff
@@ -185,6 +205,51 @@ class ArrayField extends Component {
     autofocus: false,
   };
 
+  constructor(props) {
+    super(props);
+    const { formData } = props;
+    const keyedFormData = generateKeyedFormData(formData);
+    this.state = {
+      keyedFormData,
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const nextFormData = nextProps.formData;
+    const previousKeyedFormData = this.state.keyedFormData;
+    const newKeyedFormData =
+      nextFormData.length === previousKeyedFormData.length
+        ? previousKeyedFormData.map((previousKeyedFormDatum, index) => {
+            return {
+              key: previousKeyedFormDatum.key,
+              item: nextFormData[index],
+            };
+          })
+        : generateKeyedFormData(nextFormData);
+    this.setState({
+      keyedFormData: newKeyedFormData,
+    });
+  }
+
+  /*
+  // React 16.3 replaces componentWillReceiveProps with getDerivedStateFromProps
+  //
+  static getDerivedStateFromProps(nextProps, prevState) {
```

Comment:
I think it might be cleaner (and easier to migrate to React 16) just to use getDerivedStateFromProps and https://github.com/reactjs/react-lifecycles-compat for now?

Reply:
Done in latest commit.